sound/soc/sof/topology.c

```diff
@@ -894,6 +905,65 @@ static int sof_parse_string_tokens(struct snd_soc_component *scomp,
 	return found;
 }
 
+/* this should replace sof_parse_string_tokens() after switch all token parsing */
+static int sof_parse_string_tokens_new(struct snd_soc_component *scomp,
+				   const struct sof_topology_token *tokens,
+				   int count, struct snd_soc_tplg_vendor_array *array,
+				   struct snd_sof_tuple **tuples, int *num_tuples)
+{
+	struct snd_soc_tplg_vendor_string_elem *elem;
+	int found = 0;
+	int i, j;
+	u32 value;
+
+	/* parse element by element */
+	for (i = 0; i < le32_to_cpu(array->num_elems); i++) {
+		elem = &array->string[i];
+
+		/* search for token */
+		for (j = 0; j < count; j++) {
+			struct snd_sof_tuple *new_tuples;
+
+			/* match token type */
+			if (tokens[j].type != SND_SOC_TPLG_TUPLE_TYPE_STRING)
+				continue;
+
+			/* match token id */
+			if (tokens[j].token != le32_to_cpu(elem->token))
+				continue;
+
+			/* matched - get value */
+			value = tokens[j].get_token(elem, NULL, 0, tokens[j].size);
```

Comment:
This is going to dereference badly at or just above NULL...

Reply:
No, it does not as it is going to use `find_format()` which ignores the object, offset and size